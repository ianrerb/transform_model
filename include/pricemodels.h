#ifndef PRICEMODELS_H
#define PRICEMODELS_H
#define PI 3.14159265358979
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>

class transform_model_base {//ABC
  public:
    transform_model_base(const unsigned int N_, 
			 const double alpha_, 
			 const double eta_): 
			 N(N_), alpha(alpha_), eta(eta_){ };
    virtual ~transform_model_base(){ };
    virtual double complex logCF(const double &S, const double complex &u) const=0; //pure virtual function
  protected:
    double alpha, eta;
    unsigned int N;
  };

class fft_base : public transform_model_base {//abc
  public:
    fft_base(const unsigned int N_, const double alpha_, const double eta_): transform_model_base(N_,alpha_,eta_){ };
    virtual ~fft_base(){ };
    double getLambda(){ return 2.0*PI/(static_cast<double>(N)*eta);};
    virtual double complex logCF(const double &S, const double complex &u) const=0; //pure virtual function
    std::vector<double> PutPrices(const double &S, const double &C, const double &offset=0);
  private:
    void ComputeX(double complex X[], const double &S, const double &C, const double &offset); 
  };

class GBM_model : public fft_base {
  public:
    GBM_model(const double sigma_, 
	     const double r_, 
	     const double q_, 
	     const double t_, 
	     const unsigned int N_, 
	     const double alpha_, 
	     const double eta_): 
	     fft_base(N_,alpha_,eta_),sigma(sigma_), r(r_), q(q_), t(t_) { }; 
    ~GBM_model(){ }; 
    double complex logCF(const double &S, const double complex &u) const;
  private:
    double sigma, r, q, t, alpha, eta;
    unsigned int N;
  };


#endif
