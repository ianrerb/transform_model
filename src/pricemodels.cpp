#include"../include/pricemodels.h"

void fft_base::ComputeX(double complex X[], const double &S, const double &C, const double &offset){
  double s = log(S);
  double lambda = getLambda();

  for(int j = 0; j!=N; j++){
    double complex nu = j*static_cast<double complex>(N); //nu = (j-1)eta
    double complex val1 = alpha+I*nu; //denominator of C term
    double complex temp = eta*C/(val1*(1.0+val1))*cexp(-I*(s-PI/eta))*logCF(S,nu-(alpha+1)*I); //x_j j!=0
    if(j==0){ temp*=.5; }//adjustment for j = 0
    X[j]=temp;
    }
  }

std::vector<double> fft_base::PutPrices(const double &S, const double &C, const double &offset){
  double complex xvals[N];
  printf("Computing X...\n"); 
  fft_base::ComputeX(xvals,S,C,offset);
  double complex yvals[N];
  fftw_plan p = fftw_plan_dft_1d(N,xvals,yvals,FFTW_FORWARD, FFTW_ESTIMATE);
  std::vector<double> retvals;
  printf("Computing FFT...\n"); 
  fftw_execute(p);
//  fftw_destroy_plan(p);
  for(int i =0; i!= N; i++)
    retvals.push_back(exp(-alpha*(log(S)-static_cast<double>(N)*getLambda()/2.0))*creal(yvals[i])/PI);
  return retvals;
  
  }

double complex GBM_model::logCF(const double &S, const double complex &u) const{
  double complex retval = I*(clog(S) + (r - q - cpow(sigma,2.0)/2.0)*t)*u - cpow(sigma*u,2.0)*t/2.0;
  return cexp(retval);
  }
