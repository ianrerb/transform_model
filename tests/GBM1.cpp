#include"../include/pricemodels.h"
#include<cmath>
#include<iostream>

using namespace std;
void printComplex(double complex c){
  cout<<"("<<creal(c)<<","<<cimag(c)<<")"<<endl;  
  }

int main(){
  GBM_model gbm(1,.5,0,1);
  double complex val;
  val = gbm.logCF(1,1); //should give real val of ~.6065
  printComplex(val);
   
  val = gbm.logCF(100,0); //should give real val of 1
  printComplex(val);
  
  val = gbm.logCF(exp(1),1); //should give ~.3277 + .5104i
  printComplex(val);
  return 0;
  }
