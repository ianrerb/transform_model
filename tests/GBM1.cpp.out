(0.606531,0)
(1,0)
(0.32771,0.510378)
